src/hotspot/share/gc/shared/classUnloadingContext.hpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_SHARED_CLASSUNLOADINGCONTEXT_HPP
+#define SHARE_GC_SHARED_CLASSUNLOADINGCONTEXT_HPP
+
+#include "utilities/growableArray.hpp"
+
+class ClassLoaderData;
+class Klass;
+class nmethod;
+
+class ClassUnloadingContext : public CHeapObj<mtGC> {
+protected:
+  static ClassUnloadingContext* _context;
+
+  ClassUnloadingContext();
+  virtual ~ClassUnloadingContext();
+
+public:
+  static ClassUnloadingContext* context() { assert(_context != nullptr, "context not set"); return _context; }
+
+  // Has class unloading occurred?
+  virtual bool has_unloaded_classes() const = 0;
+
+  virtual void register_unloading_class_loader_data(ClassLoaderData* cld) = 0;
+  virtual void purge_class_loader_data() = 0;
+
+  // Apply f on all Klasses of all unloading ClassLoaderDatas.
+  virtual void classes_unloading_do(void f(Klass* const)) = 0;
+
+  virtual void register_unlinked_nmethod(nmethod* nm) = 0;
+  virtual void purge_nmethods() = 0;
+  virtual void free_code_blobs() = 0;
+
+  void purge_and_free_nmethods() {
+    purge_nmethods();
+    free_code_blobs();
+  }
+};
+
+class DefaultClassUnloadingContext : public ClassUnloadingContext {
```

Comment:
I don't understand why they need to be two classes, even after reading "These are the reason for the class hierarchy for...". The reference to future/other PR(s) in the description doesn't really help -- it's unclear what is *necessary* for the current PR and what is preparation for future PR(s).

Reply:
As explained in the now removed original description:


>These are the reason for the class hierarchy for `ClassUnloadingContext`: the goal is to ultimately have about this phasing (for G1):
>1. collect all dead CLDs, using the `register_unloading_class_loader_data` method *only*
>2. parallelize the stuff in `ClassLoaderData::unload()` in one way or another, adding them to the `complete_cleaning` >(parallel) phase.
>
>Particularly the split of `SystemDictionary::do_unloading` into "only" traversing the CLDs to find the dead ones and then in parallel process them in 2. above warrants a separate `ClassUnloadingContext` (to facilitate parallelism). 


I.e. the non-parallelized case does not need the necessary data structure complications and helper methods to do efficient parallel iteration. However as mentioned I removed the class hierarchy for this change as it's unnecessary for now; let's discuss this hierarchy separately.